src/d3d12/nvapi_d3d12_device.cpp

```diff
@@ -105,17 +148,45 @@ namespace dxvk {
         return deviceExt;
     }
 
-    Com<ID3D12GraphicsCommandListExt> NvapiD3d12Device::GetCommandListExt(ID3D12GraphicsCommandList* commandList) {
+    d3d12_graphics_command_list_ext_iface* NvapiD3d12Device::GetCommandListExt(ID3D12GraphicsCommandList* commandList, uint32_t* interfaceVersion) {
         std::scoped_lock lock(m_CommandListMutex);
         auto it = m_CommandListMap.find(commandList);
-        if (it != m_CommandListMap.end())
-            return it->second;
-
-        Com<ID3D12GraphicsCommandListExt> commandListExt;
-        if (FAILED(commandList->QueryInterface(IID_PPV_ARGS(&commandListExt))))
+        if (it != m_CommandListMap.end()) {
+            *interfaceVersion = it->second.second;
+            return it->second.first;
+        }
+
+        d3d12_graphics_command_list_ext_iface* iface = nullptr;
+
+        // Grab the highest interface possible
+        do {
+            ID3D12GraphicsCommandListExt1* commandListExt1 = nullptr;
+            commandList->QueryInterface(IID_PPV_ARGS(&commandListExt1));
+
+            if (commandListExt1) {
+                *interfaceVersion = 1;
+                iface = (d3d12_graphics_command_list_ext_iface*)commandListExt1;
+                break;
+            }
+
+            ID3D12GraphicsCommandListExt* commandListExt = nullptr;
+            commandList->QueryInterface(IID_PPV_ARGS(&commandListExt));
+
+            if (commandListExt) {
+                *interfaceVersion = 0;
+                iface = (d3d12_graphics_command_list_ext_iface*)commandListExt;
+                break;
+            }
+
+            log::write("commandList has no ID3D12GraphicsCommandListExt compatible interface!");
             return nullptr;
+        } while (0);
+
+        // Release the reference that was taken by our successful QueryInterface() call
+        iface->Release();
```

Comment:
I am by no means a COM expert but this looks like it would drop the refcount back to 0 and free the object before we have a chance to use it?

Reply:
From my local experiments with the test app (and a real app) that's not the case.

Consider the code before any changes. When `LaunchCubinShader` is called we first try and lookup the `commandList` that was passed in to see if we've seen it before. If we haven't we use `QueryInterface` to populate  `Com<ID3D12GraphicsCommandListExr>` object with it's type.

At this point, the object's refcount raises from 0 to 1.

Next we extract the raw pointer from the `Com<ID3D12GraphicsCommandListExt>` object. `Com<>::ptr()` does *not* alter the refcount. This extract is done so that we can add that extracted pointer to `m_CommandListMap`.

After this we return from the function. Because `Com<ID3D12GraphicsCommandListExt>` was stack declared, it "pops" and its destructor is called. The destructor decrements the reference count for the object from 1 to 0.

Subsequent calls to `GetCommandListExt()` simply return the raw pointer value and ignore refcounting completely.

This is kind of necessary, because if we held onto a reference the whole time we'd hit issues where an application couldn't properly release it's command lists to free them because we held a reference to them.

Reply:
ah yeah, looking again at the old code (and at other cache maps), we always neutralize the ref counting from our side when putting things in the cache maps. Strict;y speaking in terms of COM, we shouldn't do that and release when we are sure that those pointers aren't used anymore (thus by e.g. storing `Com<>` in the maps instead of the raw pointers). But that would mean that we keep all those instances alive until someone calls `NvAPI_Uninitialize`. Being afraid that we keep things unnecessary alive, like Liam also stated, we went in the past for the currently implemented approach.
Currently we thus rely on good behavior of consumers and (at least to my knowledge), it hasn't exploded.